File: my_py_pkg/my_py_pkg/add_two_ints_client.py
import rclpy 
from rclpy import Future
from rclpy.node import Node 
from example_interfaces.srv import AddTwoInts 
from functools import partial


class AddTwoIntsClient(Node): 
    def __init__(self): 
        super().__init__("add_two_ints_client")
        self.call_add_two_int_server(3, 7)


    def call_add_two_int_server(self, a : int, b : int): 
        client = self.create_client(AddTwoInts, "add_two_ints")
        
        # Wait until service is up
        while not client.wait_for_service(1.0): 
            self.get_logger().warn("Waiting for server")

        # Create request
        request = AddTwoInts.Request()
        request.a = a
        request.b = b   

        future = client.call_async(request) 
        future.add_done_callback(
            partial(self.callback_add_two_ints, a=a, b=b)
        )

    def callback_add_two_ints(self, future : Future, a : int, b : int): 
        # Get response 
        try: 
            response = future.result()
            self.get_logger().info(f"Response : {a} + {b} = {response.sum}")
        except Exception as e: 
            self.get_logger().error(f"Failed {e}")

def main(args=None): 
    rclpy.init(args=args) # start ROS2 communication
    node = AddTwoIntsClient()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__": 
    main()

File: my_cpp_pkg/src/add_two_ints_client.cpp
#include <rclcpp/rclcpp.hpp>
#include <example_interfaces/srv/add_two_ints.hpp> 



class AddTwoIntsClient : public rclcpp::Node 
{
    public: 
        AddTwoIntsClient() : Node("add_two_ints_client")
        { 
            thread_.push_back(std::thread(std::bind(&AddTwoIntsClient::CallAddTwoInts, this, 1, 4)));
            thread_.push_back(std::thread(std::bind(&AddTwoIntsClient::CallAddTwoInts, this, 2, 5)));
        };

        void CallAddTwoInts(int a, int b)
        {
            auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");

            while (!client->wait_for_service(std::chrono::seconds(1)))
            {
                RCLCPP_WARN(this->get_logger(), "server is not started, waiting for server");
            } 

            auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
            request->a = 1; 
            request->b = 2;

            auto future = client->async_send_request(request); 
            try 
            { 
                auto response = future.get();
                RCLCPP_INFO(this->get_logger(), "%d + %d = %ld", a, b, response->sum);
            }
            catch (const std::exception &e)
            {
                RCLCPP_ERROR(this->get_logger(), "service call failed");
            }


        }

        private: 
            std::vector<std::thread> thread_;
};


int main(int argc, char** argv) 
{   
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClient>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
  
}